File: leetcode3/황은지/576. Out of Boundary Paths.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number} maxMove
 * @param {number} startRow
 * @param {number} startColumn
 * @return {number}
 */
var findPaths = function (m, n, maxMove, startRow, startColumn) {
  const queue = [[startRow, startColumn]];
  let head = 0;
  let count = 0;
  const dir = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  for (let i = 0; i < maxMove; i++) {
    const size = queue.length - head;
    for (let k = 0; k < size; k++) {
      const [currR, currC] = queue[head++];
      for (let j = 0; j < 4; j++) {
        const nextR = currR + dir[j][0];
        const nextC = currC + dir[j][1];
        if (nextR < 0 || nextR >= m || nextC < 0 || nextC >= n) count++;
        else queue.push([nextR, nextC]);
      }
    }
  }

  return count % (Math.pow(10, 9) + 7);
};

File: leetcode3/염혜정/576. Out of Boundary Paths.java
// bfs + dp

class Solution {
    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        final int MOD = 1_000_000_007;
        long[][] dp = new long[m][n];
        dp[startRow][startColumn] = 1;
        int count = 0;
        int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        for (int move = 0; move < maxMove; move++) {
            long[][] next = new long[m][n];
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (dp[i][j] == 0) continue;
                    for (int[] d : dirs) {
                        int ni = i + d[0];
                        int nj = j + d[1];
                        if (ni < 0 || ni >= m || nj < 0 || nj >= n) {
                            count = (int) ((count + dp[i][j]) % MOD);
                        } else {
                            next[ni][nj] = (next[ni][nj] + dp[i][j]) % MOD;
                        }
                    }
                }
            }
            dp = next;
        }

        return count;
    }
}
